### PR TITLE
Feature/rm 7603 create and setup type script projects

### PR DESCRIPTION
**Pull request by a-ctor**

### CODE REVIEW

**Comment by a-ctor** [comment]
@MichaelKetting the only thing left to do is the build dependencies between the projects. Do you want them as simple project references?

**Comment by MichaelKetting** [comment]
> 
> 
> @MichaelKetting the only thing left to do is the build dependencies between the projects. Do you want them as simple project references?

@a-ctor looked up how we did this before with typescript and no, just put them in the project.pros and and do a build dependency in the solution. It's possible we can do an actual project dependency when we switch to CPS style projects, but for now, this would make things more complicated. At least, I think that was the reason we didn't do the project reference before.

File: SecurityManager/Clients.Web.Test/Global.asax.cs [hunk 1]
```diff
@@ -47,12 +47,24 @@ protected void Application_Start (object sender, EventArgs e)
...
+      const string configuration = "Debug";
+#else
+      const string configuration = "Release";
+#endif
+
       _resourceVirtualPathProvider = new ResourceVirtualPathProvider (
           new[]
           {
-              new ResourcePathMapping ("Remotion.Web", @"..\..\Remotion\Web\Core\res"),
+              new ResourcePathMapping ("Remotion.Web/Html/jquery-1.6.4.js", @"..\..\Web\Core\res\Html\jquery-1.6.4.js"),
```
Comment (MichaelKetting) [inline SecurityManager/Clients.Web.Test/Global.asax.cs:59]: Probably should be a dedicated cleanup PR, but I have an idea how to remove the code duplication. Please create a task and we do this in the next sprint.

File: Build/Customizations/Projects.props [hunk 2]
```diff
@@ -195,6 +195,7 @@
     <ReleaseProjectFiles Include="$(SolutionDirectory)\Remotion\ObjectBinding\Security\ObjectBinding.Security.csproj"/>
     <ReleaseProjectFiles Include="$(SolutionDirectory)\Remotion\ObjectBinding\Validation\ObjectBinding.Validation.csproj"/>
     <ReleaseProjectFiles Include="$(SolutionDirectory)\Remotion\ObjectBinding\Web\ObjectBinding.Web.csproj"/>
+    <ReleaseProjectFiles Include="$(SolutionDirectory)\Remotion\ObjectBinding\Web.ClientScript\ObjectBinding.Web.ClientScript.csproj"/>
```
Comment (MichaelKetting) [inline Build/Customizations/Projects.props:198]: the updates to the nuspec file are missing. It is possible, that we will need dedicated files for web and objectbinding.web instead of being able to to use the shared file we have now.

**Review by MichaelKetting — CHANGES_REQUESTED** [review]

File: Remotion/Web/ClientScript/Scripts/Utilities.ts [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+// 
+// re-motion is distributed in the hope that it will be useful, 
+// but WITHOUT ANY WARRANTY; without even the implied warranty of 
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the 
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+// 
+
```
[the remark this answers is not in the document]
Comment (a-ctor) [reply Remotion/Web/ClientScript/Scripts/Utilities.ts:17]: As you like, currently it is part of the initial setup issue. I could either make a separate PR, commit, or issue.

File: Remotion/Web/ClientScript/Scripts/Utilities.ts [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+  };
+  
+  public static IsNull (value: unknown): value is null
+  {
+    return TypeUtility.IsDefined(value) && value == null;
+  };
+}
+
+
+class StringUtility
```
[the remark this answers is not in the document]
Comment (a-ctor) [reply Remotion/Web/ClientScript/Scripts/Utilities.ts:72]: This would be part of the second round cleanup done in RM-7630

File: Remotion/Web/ClientScript/Scripts/typings/common.d.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+// 
+// re-motion is distributed in the hope that it will be useful, 
+// but WITHOUT ANY WARRANTY; without even the implied warranty of 
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the 
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+// 
+type Dictionary<T> = { [key: string]: T | undefined };
```
[the remark this answers is not in the document]
Comment (a-ctor) [reply Remotion/Web/ClientScript/Scripts/typings/common.d.ts:17]: yes we can share this file, compiled js files have not connection to the d.ts file - as such there cannot be any name clashes since those names are not compiled into the JavaScript (as they are just types)

File: Remotion/Web/ClientScript/Scripts/Utilities.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (MichaelKetting) [reply Remotion/Web/ClientScript/Scripts/Utilities.ts:17]: separate issue please. we could this then merge with https://github.com/re-motion/Framework/pull/29#discussion_r580182753

[the remark this answers is not in the document]
Comment (a-ctor) [reply Remotion/Web/ClientScript/Scripts/Utilities.ts:17]: For now the second round cleanup I mentionend in https://github.com/re-motion/Framework/pull/29#discussion_r580182753 would be done after all the JavaScript files have been migrated while the migration of the utility files must be done immediately after this task for the other tasks to be able to start.

[the remark this answers is not in the document]
Comment (MichaelKetting) [reply Remotion/Web/ClientScript/Scripts/Utilities.ts:17]: okay (i think)

File: Remotion/ObjectBinding/Web/ObjectBinding.Web.nuspec [hunk 6]
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0"?>
+<package>
+  <metadata>
+    <id>Remotion.ObjectBinding.Web</id>
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Web/ObjectBinding.Web.nuspec:4]: Please do not fill in anything that's not already filled in in the template.

File: Remotion/Web/Core/Web.Core.nuspec [hunk 7]
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0"?>
+<package>
+  <metadata>
+    <id>Remotion.Web</id>
```
Comment (MichaelKetting) [inline Remotion/Web/Core/Web.Core.nuspec:4]: see above

**Review by MichaelKetting — CHANGES_REQUESTED** [review]

File: Remotion/ObjectBinding/Web/ObjectBinding.Web.nuspec [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    <file src="bin\$configuration$\**\$id$.resources.dll" target="lib\net461" />
+    <file src="FacetManifest.xml*" />
+    <file src="$documentationFile$" target="doc" />
+    <!-- By using a wildcard, the file-inclusion does not fail if no schema folder exists. -->
+    <file src="Schemas*\*.xsd" target="schema" />
+    <file src="..\Web.ClientScript\bin\$configuration$\dist\*" target="res\Html" />
+    <file src="Res\**\*" target="res" />
+    <file src="UI\**\*.as?x;UI\**\*.master" target="res/UI" />
+    <file src="UrlMapping.xml*" target="res" />
+  </files>
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Web/ObjectBinding.Web.nuspec:31]: There's files missing.

File: Remotion/Web/Core/Web.Core.nuspec [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    <file src="bin\$configuration$\**\$id$.resources.dll" target="lib\net461" />
+    <file src="FacetManifest.xml*" />
+    <file src="$documentationFile$" target="doc" />
+    <!-- By using a wildcard, the file-inclusion does not fail if no schema folder exists. -->
+    <file src="Schemas*\*.xsd" target="schema" />
+    <file src="..\ClientScript\bin\$configuration$\dist\*" target="res\Html" />
+    <file src="Res\**\*" target="res" />
+    <file src="UI\**\*.as?x;UI\**\*.master" target="res/UI" />
+    <file src="UrlMapping.xml*" target="res" />
+  </files>
```
Comment (MichaelKetting) [inline Remotion/Web/Core/Web.Core.nuspec:31]: see above

**Review by MichaelKetting — CHANGES_REQUESTED** [review]

**Review by MichaelKetting — COMMENTED** [review]
Looks good. will check the nuget packages after the nightly build is done.

**Comment by a-ctor** [comment]
Squashed the review changes and rebased on develop so that we can merge in the morning if the nightly artifacts are correct.

File: Build/Customizations/Projects.props (hunk 2 above)
Comment (MichaelKetting) [inline Build/Customizations/Projects.props:198]: should put the Clientscript references first. right now, we don't have a dependency (i.e. building the nuget pacakge immediately after compiling a project, but that's not guaranteed.

File: Remotion/Development/Web/ResourceHosting/FileExtensionHandlerMapping.cs [hunk 10]
```diff
@@ -36,6 +36,7 @@ public static FileExtensionHandlerMapping[] Default
                    new FileExtensionHandlerMapping ("png", ResourceVirtualPathProvider.StaticFileHandler),
                    new FileExtensionHandlerMapping ("gif", ResourceVirtualPathProvider.StaticFileHandler),
                    new FileExtensionHandlerMapping ("js", ResourceVirtualPathProvider.StaticFileHandler),
+                   new FileExtensionHandlerMapping ("map", ResourceVirtualPathProvider.StaticFileHandler),
```
Comment (MichaelKetting) [inline Remotion/Development/Web/ResourceHosting/FileExtensionHandlerMapping.cs:39]: that change should be in the last commit since we don't integrate the TS files before then.

File: Remotion.sln [hunk 11]
```diff
@@ -347,6 +353,10 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "ObjectBinding.Web.Compatibi
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Build", "Build\Build.csproj", "{6CBB099A-DE58-48A4-ADC7-D37725B0181C}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Web.ClientScript", "Remotion\Web\ClientScript\Web.ClientScript.csproj", "{627D9351-BE97-485E-A720-36C732056FF8}"
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "ObjectBinding.Web.ClientScript", "Remotion\ObjectBinding\Web.ClientScript\ObjectBinding.Web.ClientScript.csproj", "{2DA25FE6-18AD-4E69-B860-F30346965646}"
```
Comment (MichaelKetting) [inline Remotion.sln:358]: there's no project build order dependency for ObjectBinding.Web.ClientScript on ObjectBinding.Web

File: .gitignore [hunk 12]
```diff
@@ -194,3 +194,7 @@ GitExtensions.settings.backup
 
 # re-motion
 Remotion/ObjectBinding/Web.Test/log-data/
+
+# Do not ignore TypeScript build config directories
```
Comment (MichaelKetting) [inline .gitignore:198]: this would be best as it's own commit at the start of the branch

File: Remotion/Web/Core/Web.Core.csproj [hunk 13]
```diff
@@ -876,16 +873,8 @@
     </None>
     <None Include="Res\HTML\jquery-1.6.4.js" />
     <None Include="Res\HTML\jquery.iFrameShim.js" />
-    <None Include="Res\HTML\DatePicker.js" />
```
Comment (MichaelKetting) [inline Remotion/Web/Core/Web.Core.csproj:879]: dropping the JS files, probably best as a last, separate commit instead of the "Register output" commit. in theory, the best place would be when we move the files, but that's just too complicated if we want to try for "always runnable"

**Review by MichaelKetting — CHANGES_REQUESTED** [review]

**Review by MichaelKetting — APPROVED** [review]
Update looks good.
